File: apps/jump-ios/YPJump/xLENS/DeviceKeyManager.swift
// ═══════════════════════════════════════════════════════════════
// DEVICE KEY MANAGER
// Hardware-backed signing keys using Secure Enclave
// Layer 1 + 2 of xLENS Three-Layer Defense
// ═══════════════════════════════════════════════════════════════

import Foundation
import Security
import CryptoKit

/// Manages hardware-backed device keys for xLENS verification
/// Uses iOS Secure Enclave for tamper-resistant key storage
@MainActor
final class DeviceKeyManager: ObservableObject {

    // MARK: - Published State

    @Published private(set) var deviceKey: DeviceKey?
    @Published private(set) var isRegistered = false
    @Published private(set) var error: XLENSError?

    // MARK: - Private Properties

    private let keyTag = "com.ypjump.xlens.devicekey"
    private let keychainService = "YPJumpXLENS"

    // MARK: - Key Generation

    /// Generate a new device key using Secure Enclave
    /// Returns the key ID if successful
    func generateDeviceKey() async throws -> String {
        // Delete any existing key
        deleteExistingKey()

        // Create key with Secure Enclave protection
        let access = SecAccessControlCreateWithFlags(
            kCFAllocatorDefault,
            kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
            [.privateKeyUsage],
            nil
        )!

        var attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrKeySizeInBits as String: 256,
            kSecAttrTokenID as String: kSecAttrTokenIDSecureEnclave,
            kSecPrivateKeyAttrs as String: [
                kSecAttrIsPermanent as String: true,
                kSecAttrApplicationTag as String: keyTag.data(using: .utf8)!,
                kSecAttrAccessControl as String: access
            ]
        ]

        var error: Unmanaged<CFError>?
        guard let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, &error) else {
            // Fallback to software key if Secure Enclave not available (e.g., simulator)
            print("⚠️ DeviceKeyManager: Secure Enclave not available, using software key")
            return try generateSoftwareKey()
        }

        // Get public key
        guard let publicKey = SecKeyCopyPublicKey(privateKey) else {
            throw XLENSError.signatureFailed
        }

        // Export public key as raw data
        guard let publicKeyData = SecKeyCopyExternalRepresentation(publicKey, nil) as Data? else {
            throw XLENSError.signatureFailed
        }

        let keyId = generateKeyId()
        let publicKeyBase64 = publicKeyData.base64EncodedString()

        deviceKey = DeviceKey(
            keyId: keyId,
            publicKey: publicKeyBase64,
            platform: "ios",
            deviceModel: deviceModel,
            osVersion: osVersion,
            hardwareLevel: .strongbox
        )

        isRegistered = true

        print("✅ DeviceKeyManager: Generated Secure Enclave key")
        print("   Key ID: \(keyId)")

        return keyId
    }

    /// Fallback: Generate software-only key (for simulator)
    private func generateSoftwareKey() throws -> String {
        let privateKey = P256.Signing.PrivateKey()
        let publicKeyData = privateKey.publicKey.rawRepresentation

        // Store in keychain
        let keyId = generateKeyId()

        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: keyTag.data(using: .utf8)!,
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecValueData as String: privateKey.rawRepresentation,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]

        let status = SecItemAdd(query as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw XLENSError.signatureFailed
        }

        deviceKey = DeviceKey(
            keyId: keyId,
            publicKey: publicKeyData.base64EncodedString(),
            platform: "ios",
            deviceModel: deviceModel,
            osVersion: osVersion,
            hardwareLevel: .software
        )

        isRegistered = true

        print("✅ DeviceKeyManager: Generated software key (Secure Enclave not available)")
        print("   Key ID: \(keyId)")

        return keyId
    }

    // MARK: - Signing

    /// Sign data using the device key
    /// Returns base64-encoded ES256 signature
    func sign(data: Data) async throws -> String {
        // Try Secure Enclave key first
        if let signature = try? signWithSecureEnclave(data: data) {
            return signature
        }

        // Fallback to software key
        return try signWithSoftwareKey(data: data)
    }

    private func signWithSecureEnclave(data: Data) throws -> String {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: keyTag.data(using: .utf8)!,
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecReturnRef as String: true
        ]

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)

        guard status == errSecSuccess, let privateKey = item else {
            throw XLENSError.deviceKeyNotFound
        }

        let key = privateKey as! SecKey

        var error: Unmanaged<CFError>?
        guard let signature = SecKeyCreateSignature(
            key,
            .ecdsaSignatureMessageX962SHA256,
            data as CFData,
            &error
        ) as Data? else {
            throw XLENSError.signatureFailed
        }

        return signature.base64EncodedString()
    }

    private func signWithSoftwareKey(data: Data) throws -> String {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: keyTag.data(using: .utf8)!,
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecReturnData as String: true
        ]

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)

        guard status == errSecSuccess, let keyData = item as? Data else {
            throw XLENSError.deviceKeyNotFound
        }

        let privateKey = try P256.Signing.PrivateKey(rawRepresentation: keyData)
        let signature = try privateKey.signature(for: data)

        return signature.rawRepresentation.base64EncodedString()
    }

    // MARK: - Key Management

    /// Check if device has a registered key
    func checkRegistration() -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: keyTag.data(using: .utf8)!,
            kSecReturnRef as String: true
        ]

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)

        isRegistered = status == errSecSuccess
        return isRegistered
    }

    /// Delete existing device key
    func deleteExistingKey() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: keyTag.data(using: .utf8)!
        ]

        SecItemDelete(query as CFDictionary)
        isRegistered = false
        deviceKey = nil
    }

    /// Load stored key ID from keychain
    func loadStoredKeyId() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: "keyId",
            kSecReturnData as String: true
        ]

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)

        guard status == errSecSuccess, let data = item as? Data else {
            return nil
        }

        return String(data: data, encoding: .utf8)
    }

    /// Store key ID in keychain
    private func storeKeyId(_ keyId: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: "keyId",
            kSecValueData as String: keyId.data(using: .utf8)!
        ]

        // Delete existing
        SecItemDelete(query as CFDictionary)

        // Add new
        SecItemAdd(query as CFDictionary, nil)
    }

    // MARK: - Helpers

    private func generateKeyId() -> String {
        var bytes = [UInt8](repeating: 0, count: 12)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        let base64 = Data(bytes).base64EncodedString()
            .replacingOccurrences(of: "+", with: "")
            .replacingOccurrences(of: "/", with: "")
            .replacingOccurrences(of: "=", with: "")
        return "dk_\(base64)"
    }

    private var deviceModel: String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        return identifier
    }

    private var osVersion: String {
        "\(ProcessInfo.processInfo.operatingSystemVersion.majorVersion).\(ProcessInfo.processInfo.operatingSystemVersion.minorVersion).\(ProcessInfo.processInfo.operatingSystemVersion.patchVersion)"
    }
}

File: apps/jump-android/app/src/main/java/com/youthperformance/jump/xlens/DeviceKeyManager.kt
// ═══════════════════════════════════════════════════════════════
// DEVICE KEY MANAGER
// Hardware-backed signing keys using Android Keystore/StrongBox
// Layer 1 + 2 of xLENS Three-Layer Defense
// ═══════════════════════════════════════════════════════════════

package com.youthperformance.jump.xlens

import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.PrivateKey
import java.security.SecureRandom
import java.security.Signature
import java.security.spec.ECGenParameterSpec

/**
 * Manages hardware-backed device keys for xLENS verification
 * Uses Android Keystore with StrongBox when available
 */
class DeviceKeyManager {

    companion object {
        private const val TAG = "DeviceKeyManager"
        private const val KEY_ALIAS = "ypjump_xlens_device_key"
        private const val KEYSTORE_PROVIDER = "AndroidKeyStore"
    }

    // State
    private val _deviceKey = MutableStateFlow<DeviceKey?>(null)
    val deviceKey: StateFlow<DeviceKey?> = _deviceKey.asStateFlow()

    private val _isRegistered = MutableStateFlow(false)
    val isRegistered: StateFlow<Boolean> = _isRegistered.asStateFlow()

    private val _error = MutableStateFlow<XLENSError?>(null)
    val error: StateFlow<XLENSError?> = _error.asStateFlow()

    private val keyStore: KeyStore = KeyStore.getInstance(KEYSTORE_PROVIDER).apply {
        load(null)
    }

    /**
     * Generate a new device key using Android Keystore
     * Returns the key ID if successful
     */
    suspend fun generateDeviceKey(): String {
        // Delete any existing key
        deleteExistingKey()

        val keyId = generateKeyId()
        val hardwareLevel = determineHardwareLevel()

        try {
            // Build key generation parameters
            val keyGenBuilder = KeyGenParameterSpec.Builder(
                KEY_ALIAS,
                KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY
            )
                .setDigests(KeyProperties.DIGEST_SHA256)
                .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
                .setUserAuthenticationRequired(false)

            // Try to use StrongBox if available (Android 9+)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && hardwareLevel == DeviceKey.HardwareLevel.STRONGBOX) {
                try {
                    keyGenBuilder.setIsStrongBoxBacked(true)
                    Log.i(TAG, "Using StrongBox for key generation")
                } catch (e: Exception) {
                    Log.w(TAG, "StrongBox not available, falling back to TEE")
                }
            }

            val keyGenParameterSpec = keyGenBuilder.build()

            // Generate the key pair
            val keyPairGenerator = KeyPairGenerator.getInstance(
                KeyProperties.KEY_ALGORITHM_EC,
                KEYSTORE_PROVIDER
            )
            keyPairGenerator.initialize(keyGenParameterSpec)
            val keyPair = keyPairGenerator.generateKeyPair()

            // Get public key as Base64
            val publicKeyBase64 = Base64.encodeToString(
                keyPair.public.encoded,
                Base64.NO_WRAP
            )

            _deviceKey.value = DeviceKey(
                keyId = keyId,
                publicKey = publicKeyBase64,
                platform = "android",
                deviceModel = "${Build.MANUFACTURER}_${Build.MODEL}",
                osVersion = Build.VERSION.RELEASE,
                hardwareLevel = hardwareLevel
            )

            _isRegistered.value = true

            Log.i(TAG, "✅ Generated device key")
            Log.i(TAG, "   Key ID: $keyId")
            Log.i(TAG, "   Hardware Level: $hardwareLevel")

            return keyId

        } catch (e: Exception) {
            Log.e(TAG, "Failed to generate device key", e)
            _error.value = XLENSError.SignatureFailed
            throw e
        }
    }

    /**
     * Sign data using the device key
     * Returns base64-encoded ES256 signature
     */
    suspend fun sign(data: ByteArray): String {
        try {
            val privateKey = keyStore.getKey(KEY_ALIAS, null) as? PrivateKey
                ?: throw XLENSError.DeviceKeyNotFound

            val signature = Signature.getInstance("SHA256withECDSA")
            signature.initSign(privateKey)
            signature.update(data)

            val signatureBytes = signature.sign()
            return Base64.encodeToString(signatureBytes, Base64.NO_WRAP)

        } catch (e: XLENSError) {
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "Failed to sign data", e)
            throw XLENSError.SignatureFailed
        }
    }

    /**
     * Check if device has a registered key
     */
    fun checkRegistration(): Boolean {
        val hasKey = keyStore.containsAlias(KEY_ALIAS)
        _isRegistered.value = hasKey
        return hasKey
    }

    /**
     * Delete existing device key
     */
    fun deleteExistingKey() {
        if (keyStore.containsAlias(KEY_ALIAS)) {
            keyStore.deleteEntry(KEY_ALIAS)
        }
        _isRegistered.value = false
        _deviceKey.value = null
    }

    /**
     * Determine the hardware security level available
     */
    private fun determineHardwareLevel(): DeviceKey.HardwareLevel {
        return when {
            // Check for StrongBox (Android 9+)
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && hasStrongBox() -> {
                DeviceKey.HardwareLevel.STRONGBOX
            }
            // Check for TEE
            hasTee() -> {
                DeviceKey.HardwareLevel.TEE
            }
            // Software fallback
            else -> {
                DeviceKey.HardwareLevel.SOFTWARE
            }
        }
    }

    private fun hasStrongBox(): Boolean {
        return try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                // Try to create a StrongBox-backed key to check availability
                val testAlias = "ypjump_strongbox_test"
                val keyGenParameterSpec = KeyGenParameterSpec.Builder(
                    testAlias,
                    KeyProperties.PURPOSE_SIGN
                )
                    .setDigests(KeyProperties.DIGEST_SHA256)
                    .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
                    .setIsStrongBoxBacked(true)
                    .build()

                val keyPairGenerator = KeyPairGenerator.getInstance(
                    KeyProperties.KEY_ALGORITHM_EC,
                    KEYSTORE_PROVIDER
                )
                keyPairGenerator.initialize(keyGenParameterSpec)
                keyPairGenerator.generateKeyPair()

                // Clean up test key
                keyStore.deleteEntry(testAlias)
                true
            } else {
                false
            }
        } catch (e: Exception) {
            false
        }
    }

    private fun hasTee(): Boolean {
        // Most modern Android devices have TEE
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
    }

    private fun generateKeyId(): String {
        val bytes = ByteArray(12)
        SecureRandom().nextBytes(bytes)
        val base64 = Base64.encodeToString(bytes, Base64.NO_WRAP or Base64.URL_SAFE)
            .replace("+", "")
            .replace("/", "")
            .replace("=", "")
        return "dk_$base64"
    }
}
